service/qos/service_level_controller.cc

```diff
@@ -230,16 +230,41 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
     });
 }
 
-future<>  service_level_controller::notify_service_level_added(sstring name, service_level sl_data) {
-    _service_levels_db.emplace(name, sl_data);
-    return make_ready_future();
+future<>  service_level_controller::notify_service_level_added(sstring name, service_level sl_data) {    
+    return parallel_for_each(_subscribers, [name, sl_data] (subscriber_registry_entry& e) {
+        if (e->_qccs) {
+            return e->_qccs->on_before_service_level_add(name, sl_data.slo);
+        } else {            
+            return make_ready_future();
+        }
+    }).then([this, name, sl_data] () {
+        std::remove_if(_subscribers.begin(), _subscribers.end(), [] (subscriber_registry_entry& e) {
+            return *e == (qos_configuration_change_subscriber*)nullptr;
+        });
+        _service_levels_db.emplace(name, sl_data);
+        return make_ready_future();
+    });
+    
 }
 
 future<> service_level_controller::notify_service_level_updated(sstring name, service_level_options slo) {
     auto sl_it = _service_levels_db.find(name);
     future<> f = make_ready_future();
     if (sl_it != _service_levels_db.end()) {
-        sl_it->second.slo = slo;
+        service_level_options slo_before = sl_it->second.slo;
```

Comment:
Are these notifications serialized? If not, then a later notification could execute earlier than a previous notification.

Are notifications protected against unsubscribes? We have atomic_vector (that @psarna also mentioned) to assist with this.

Reply:
The serialization is guarantied, this is guarantied even before this change.  The calls to all of the "notify*" functions are serialized.
I have amended the code to use atomic_vector.